### PR TITLE
add support for nlargest in SeriesGroupBy

### PR DESCRIPTION
**Pull request by madhur-tandon**
- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

Closes https://github.com/dask/dask/issues/6161

The test fails for a very minor reason, I could use some help there.

### CODE REVIEW

**Comment by madhur-tandon** [comment]
Hi @mrocklin @TomAugspurger 
Could you please have a look at this?

The test fails due to some `metadata` issue.

I think there is some minor modification that I am missing. Let me know, Thanks!

**Comment by madhur-tandon** [comment]
It seems like in the `_check_dask` function in `dask/dataframe/utils.py`,
`dsk.compute().index` and `dsk._meta.index` differ.

The correct value of the index of the final answer corresponds to `dsk.compute().index`.

Any reason why this could occur?

The functions for `chunk` and `aggregate` only use `pandas` operations inside them. 
I don't think there is a way I can access this `dsk` object inside those.

**Comment by TomAugspurger** [comment]
I haven't looked closely, but this is saying that the metadata for the index was not specified correctly.

There's a mismatch between `df._meta.index.names` and `df.compute().index.names`. We need to make sure that the metadata is correct.

I'm not sure how aca determines the index for meta, or whether it allows you to specify it.

**Comment by madhur-tandon** [comment]
Yes, I figured that out already. But I was wondering how to specify the metadata.

**Comment by madhur-tandon** [comment]
The default assumption for the `metadata` is a `RangeIndex` (by Dask).

However, the final answer can have either `RangeIndex` or `MultiIndex`. This is a `pandas` problem and is unrelated to `dask`. Basically, after applying `nlargest(K)` on a grouped pandas DataFrame, if the result is all of the rows i.e. each group having K rows = the input data itself, we get a `RangeIndex` through pandas.

If the above is not the case, we get a `MultiIndex`.

Now, this seems fine, but while using the `Apply-Concat-Apply` paradigm, it is possible that some blocks of data return the `RangeIndex` while some return the `MultiIndex`.

This doesn't allow us to normally concatenate different blocks.

(unlike in https://blog.dask.org/2019/10/08/df-groupby as the `sr3` block has it's output as all of the rows of input.
Thus, it is consistent with `sr1` and `sr2` in terms of retaining it's index and all `3` of them can be concatenated easily).

Coming back to `nlargest`, since the indexes of different blocks are different. That's why there is a `if-else` block in this PR in the `_nlargest_chunk` function that accounts for this, and converts the `RangeIndex` to a `MultiIndex`, so that different blocks can be concatenated.

What I am essentially trying to say is that my implementations respects the output (as `pandas` wants it) i.e. `RangeIndex` or `MultiIndex` depending on the data and the `K` in `nlargest(K)`.

However, I am not sure how to deal with this `metadata` issue. Hence, I am changing the test simply from `nlargest(1)` to `nlargest(2)`. This way, the result would indeed contain a `RangeIndex` and would thus respect the default settings for metadata as set by dask (i.e. a `RangeIndex` indeed).

**Comment by madhur-tandon** [comment]
Update: I have also sent a new commit that covers / fixes the case when a column being accessed is also present in the grouping keys (eg: `dask_df.groupby(["location", "percent"])["percent"].nlargest(2)`)  i.e. `percent` appears in both these places (playing different roles).

**Comment by madhur-tandon** [comment]
Update: This PR is now ready.
CC: @TomAugspurger @mrocklin

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -1866,6 +1891,55 @@ def unique(self, split_every=None, split_out=1):
...
+
+def _nlargest_aggregate(df, **kwargs):
+    n = kwargs.pop("n")
+    keep = kwargs.pop("keep")
+    name = kwargs.pop("name")
+    index = kwargs.pop("index")
+    num_records = kwargs.pop("len")
+    indices = list(index)
+    df = df.reset_index()
+    level_num = list(df.columns).index(name) - 1
```
Comment (TomAugspurger) [inline dask/dataframe/groupby.py:1929]: `name` is always in `df.columns`? Or can it be referring to the level of an index? Pandas (and I think Dask) supports either.

File: dask/dataframe/tests/test_groupby.py [hunk 2]
```diff
@@ -404,6 +404,18 @@ def test_groupby_get_group():
         assert_eq(ddgrouped.a.get_group(2), pdgrouped.a.get_group(2))
 
 
+def test_dataframe_groupby_nlargest():
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_groupby.py:407]: Can you parametrize this by one or two levels: `"location"` and `["location", "category"]`. There's been issues around this in the past.

**Review by TomAugspurger — COMMENTED** [review]
Thanks, a couple comments.

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (madhur-tandon) [reply dask/dataframe/groupby.py:1929]: Yes, the `name` should be present in `df.columns` always.
In the operation, `dask_df.groupby(["location", "category"])["percent"].nlargest(2)`, the name is `percent`. It represents the column for which we want these `nlargest` values.

Since this is always passed before calling `.nlargest(K)`, thus the `percent` column will be always present in the result (irrespective of if it's a `MultiIndex` or a `RangeIndex`). Now, when operations such as `to_frame()` OR perhaps `.reset_index()` are done over it, we get the column `percent` back, which was hidden before (in a `MultiIndex` or a `RangeIndex`)

Since the operation `.reset_index()` is applied before calculating this `level_num` value, we are bound to get a valid value for this variable.

This is expected because the `nlargest` function that we implemented here is for `groupby.SeriesGroupBy.nlargest` object (and not for `DataFrame.nlargest` or `Series.nlargest` -- they are already implemented in Dask) which can be only called after we access a column first (in this case the `percent` column).

File: dask/dataframe/tests/test_groupby.py (hunk 2 above)
Comment (madhur-tandon) [reply dask/dataframe/tests/test_groupby.py:407]: I don't quite understand what you mean. Can you elaborate a bit please? What does `this` in `parametrize this` mean? Do you want me to make 1 test for just `location` and a 2nd test for `["location", "category"]`?

**Comment by madhur-tandon** [comment]
There is a minor case that I just discovered. I shall try to fix it tomorrow or the day after. It's when the `_nlargest_aggregate` function applies `nlargest` on the final aggregated dataframe (the `concat` of each of the chunks) and when it's number of rows = the number of rows of the output (but still less than number of rows of the original input). Although the answer matches, there will be a discrepancy again between `RangeIndex` and `MultiIndex`.

File: dask/dataframe/tests/test_groupby.py (hunk 2 above)
Comment (madhur-tandon) [reply dask/dataframe/tests/test_groupby.py:407]: I guess you meant using `pytest.mark.parametrize`. Doing it now. Sorry, it was unclear before.

Comment (TomAugspurger) [reply dask/dataframe/tests/test_groupby.py:407]: Yep, exactly. Roughly

```python
pytest.mark.parametrize("level", ["location", ["location", "category"])
def test_dataframe_groupby_nlargest(level):
    ...
```

And then use `level` in the groupby. If it makes the test unnecessarily complex you can split it into two separate tests.

**Comment by madhur-tandon** [comment]
@TomAugspurger @mrocklin I have modified the test a bit. It is possible that although both `pandas` and `dask` return the correct results, their order might be different (due to splitting in chunks and concatenating and re-applying). Further, the final result can have a `RangeIndex`, a `MultiIndex` or even an `Int64Index`. Thus, to account for equality of the results, I have converted the outputs to dictionaries and then compared them.

I have also added a total of 24 tests in the latest commit. Hopefully, this should suffice.

File: dask/dataframe/tests/test_groupby.py (hunk 2 above)
Comment (madhur-tandon) [reply dask/dataframe/tests/test_groupby.py]: Done! A total of 24 tests have been added!

**Comment by madhur-tandon** [comment]
The tests pass for me locally. It seems they only fail for `Windows CI / build (3.6)`. Any tips on how I can reproduce locally?

**Comment by madhur-tandon** [comment]
The raw logs indicate the error as `nlargest() got an unexpected keyword argument 'axis'` which is probably an issue for an older / different version of pandas. I am switching to `keyword arguments` that will hopefully fix this.

**Comment by madhur-tandon** [comment]
Update: The total number of tests are now 36. All of them pass locally for me. More corner cases have been fixed with the latest commit.

File: dask/dataframe/tests/test_groupby.py (hunk 2 above)
Comment (madhur-tandon) [reply dask/dataframe/tests/test_groupby.py]: Update: There are 36 tests now :)

**Comment by madhur-tandon** [comment]
I have gone over 540 possible combinations of what we can pass as `fields` for groupby and the `name` for which we seek `nlargest` values. All of them pass except the case when both the field and the name are the same.

For eg: `dask_df.groupby(["percent"])["percent"].nlargest(n=n, keep=keep).compute()`

**Comment by madhur-tandon** [comment]
@TomAugspurger @mrocklin This PR should be finally ready. I have added 576 tests in total, all of which pass on my local machine. Hopefully, they cover all possible variations of inputs that can be passed. I can add more if needed.

I could now appreciate a review :)

I am just unsure if the tests will pass on `Windows CI / build (3.6)` due to an older / different version of pandas perhaps. The error that comes is `nlargest() got an unexpected keyword argument 'axis'`. I hope it passes though.

**Comment by madhur-tandon** [comment]
Update: This should be ready to merge!

**Comment by martindurant** [comment]
cc @dask/maintenance 
@madhur-tandon : note that you have failures on CI

**Comment by madhur-tandon** [comment]
Hi @martindurant Yup, but they are unrelated to this PR.
It probably occurs due to an older / different version of pandas for `Python 3.6`.

The error is `nlargest() got an unexpected keyword argument 'axis'`.

Perhaps the older version of pandas used to have this argument. But I am not able to find the relevant old docs for it.
I am certain that upgrading pandas on the `Windows CI / build (3.6)` machine should fix it.

**Comment by martindurant** [comment]
It would be better to add an xfail for the specific condition, with an explanation - otherwise all other test runs would fail after merging this

**Comment by madhur-tandon** [comment]
Alternatively, one could remove this config since `pandas is dropping support for Python 3.6` and `dask` depends on `pandas`. (https://github.com/pandas-dev/pandas/pull/35214)

**Comment by martindurant** [comment]
In a separate discussion, we will drop py36 support sometime soon, but I would xfail this anyway for now (*only* for py36)

**Comment by madhur-tandon** [comment]
Done, Hopefully everything should pass now 👍

File: dask/dataframe/tests/test_groupby.py [hunk 3]
```diff
@@ -433,8 +433,10 @@ def test_groupby_get_group():
 
 
 @pytest.mark.xfail(
-    sys.version_info[0] == 3 and sys.version_info[1] == 6,
-    reason="nlargest() behaves differently with an older version of pandas on Python 3.6 "
+    sys.version_info[0] == 3
+    and sys.version_info[1] == 6
+    and PANDAS_VERSION == "0.23.4",
```
Comment (martindurant) [inline dask/dataframe/tests/test_groupby.py:438]: This is OK, since pandas won't release any more for py36 (right, @TomAugspurger ?). Might be cleaner with a relative or LooseVersion comparison.

File: dask/dataframe/groupby.py [hunk 4]
```diff
@@ -1866,6 +1891,77 @@ def unique(self, split_every=None, split_out=1):
...
+        df.columns = [name]
+    df = df.reset_index()
+    level_num = list(df.columns).index(name) - 1
+    df.index = df["level_{0}".format(level_num)].rename(None)
+    df = df.sort_index()
+    ret = df.groupby(indices)[name].nlargest(n=n, keep=keep)
+    if (
+        type(ret.index) in [pd.RangeIndex, pd.Int64Index]
+        and len(ret) == len(df) != num_records
+    ):
```
Comment (TomAugspurger) [inline dask/dataframe/groupby.py:1954]: I'm having trouble following this section. What sorting of groupby arguments get us here?

File: dask/dataframe/groupby.py [hunk 5]
```diff
@@ -1866,6 +1891,77 @@ def unique(self, split_every=None, split_out=1):
...
+        ret = _convert_RangeIndex_to_MultiIndex(ret, df, indices, name)
+        if name in indices:
+            indices[indices.index(name)] = df[name]
+            indices.append(df.index)
+            return ret.to_frame().reset_index().set_index(indices)[name]
+    if ret.name == "temp":
+        ret.name = name_backup
```
Comment (TomAugspurger) [inline dask/dataframe/groupby.py:1961]: Is `name_backup` always going to be defined here? (Maybe use the equivalent condition as above when it's created).

File: dask/dataframe/tests/test_groupby.py [hunk 6]
```diff
@@ -404,6 +407,53 @@ def test_groupby_get_group():
         assert_eq(ddgrouped.a.get_group(2), pdgrouped.a.get_group(2))
 
 
+params = [
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_groupby.py:410]: Roughly how long to all these tests take? All else equal it's good to have more test coverage, but heavy parametrization can create very many tasks leading to slower collection / test times.

File: dask/dataframe/tests/test_groupby.py [hunk 7]
```diff
@@ -404,6 +407,53 @@ def test_groupby_get_group():
...
+@pytest.mark.parametrize("n, indices, name, keep", list(it.product(*params)))
+def test_dataframe_groupby_nlargest(n, indices, name, keep):
+    data = {}
+    data["location"] = ["A", "A", "A", "B", "C", "C", "D", "D", "D", "D", "D"]
+    data["category"] = [5, 5, 5, 3, 2, 4, 2, 3, 2, 5, 2]
+    data["percent"] = [100, 100, 100, 100, 50, 13, 75, 59, 13, 4, 75]
+    data["total"] = [45, 34, 66, 26, 72, 9, 52, 70, 60, 11, 9]
+    df = pd.DataFrame.from_dict(data)
+    dask_df = dd.from_pandas(df, npartitions=3)
+    expected = dask_df.groupby(indices)[name].nlargest(n=n, keep=keep).compute()
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_groupby.py:452]: You can leave off the `compute`. `assert_eq` will do that (and check some things about the task graph too).

Not a big deal, but typically we would call this `result` and the pandas output the `expected`.

File: dask/dataframe/groupby.py (hunk 4 above)
Comment (madhur-tandon) [reply dask/dataframe/groupby.py:1954]: If you recall one of my earliest comments, we know that pandas produces a `RangeIndex` when after applying the `groupby` operation, we get the output same as input (a necessary but insufficient way to check this is that number of rows are equal). Now, after the `groupby` operation has been applied on the chunks and their intermediate result is concatenated (let's call this `df`), it's possible that this `intermediate concatenated df` has less number of rows than the original input (`num_records`).

Now, when on this `intermediate concatenated df`, when we apply groupby once again, pandas will see this as if it received a fresh input and the result of groupby on this can lead to the final output which is same as it's input (i.e. the `intermediate concatenated df`). So we check this using `len(ret) == len(df)`.

Now, if the final output indeed had number of records equal to `num_records` (i.e. `len(ret) == len(df) == num_records`), then we want a `RangeIndex` for sure.

Otherwise, we want a `MultiIndex` but pandas gave us a `RangeIndex` (on the basis of `len(ret) == len(df)` only since it knows nothing about the original input, it has only dealt with the `intermediate concatenated df` in the `aggregation` step).

Thus, we check this condition and also check that the index is indeed of type `RangeIndex`.

In practice, since I am preserving indices at all steps, it's possible that when `len(ret) == len(df) != num_records`, i.e. some rows are missing, thus, some indices are missing, thus, it cannot be a RangeIndex (but still contains integers). So in practice, it is sufficient to check only for `pd.Int64Index`.

All the tests will still pass if I only check for `type(ret.index) == pd.Int64Index`. But to be doubly sure, I added the condition `type(ret.index) in [pd.RangeIndex, pd.Int64Index]`.

Btw, one of the parameters for the groupby operation that leads to this is: 
`n=2, indices=['total'], name='category', keep=first`

File: dask/dataframe/groupby.py (hunk 5 above)
Comment (madhur-tandon) [reply dask/dataframe/groupby.py:1961]: `name_backup` is defined whenever we have a case like:
`dask_df.groupby(["percent"])["percent"].nlargest(n=n, keep=keep).compute()`

In such a case, `pandas` won't allow operations such as `to_frame()` or perhaps `reset_index` (which are intermediate steps of this parallel `nlargest` process.)

Thus, we need to change the name to `temp` here to proceed while keeping rest of the implementation same.
Now, whenever the `name` is changed to `temp`, `name_backup` is created to back up the original value.

Since the initial event is simultaneous, thus, this condition suffices i.e. if `ret.name` is `temp`, then we did create a backup before. So `name_backup` will always exist here.

I don't think I can use the original condition here since the original condition checked for `name in indices`, but we changed `name` to `temp` afterwards.

File: dask/dataframe/tests/test_groupby.py (hunk 6 above)
Comment (madhur-tandon) [reply dask/dataframe/tests/test_groupby.py:410]: On my machine locally, all of the 576 tests take around 40 seconds.

File: dask/dataframe/tests/test_groupby.py (hunk 7 above)
Comment (madhur-tandon) [reply dask/dataframe/tests/test_groupby.py:452]: I am afraid that dropping off `compute()` will again lead us to the issue of `metadata` thingie.
The culprit here is again the weird behavior of `pandas` i.e. returning `RangeIndex` or `MultiIndex`.

So different chunks will have different kinds of indices at run time.

Btw, I checked the tests after removing `compute()`.
It turns out that many of them start failing and the point of failure seems to be in the step:
```
a = ddf._meta.dtype
b = res.dtype
> assert eq_dtypes(a, b)
```

I took the parameters of one of the failing tests, and ran only that separately with a different script. On explicit checking, these 2 turn out to be equal. So I think I don't have a clue as to why this is happening.

Using `compute()` again makes all of them pass though.

PS -- Do you want me to send an updated PR with renaming those variables as `result` and `expected`?
Let me know, thanks!

Comment (TomAugspurger) [reply dask/dataframe/tests/test_groupby.py:452]: In general, we want `result._meta` to be accurate. This is running into the issue you describe in https://github.com/dask/dask/pull/6627/files#r494596346?

Do you know, is this the same issue as https://github.com/pandas-dev/pandas/issues/34809, which https://github.com/pandas-dev/pandas/pull/34998 is attempting to fix?

**Comment by scharlottej13** [comment]
Hi @madhur-tandon! Thank you for your patience on this PR. I realize it's been a while, would you still be interested in picking this back up again?
